cmd/kubeadm/app/cmd/alpha/certs.go

```diff
@@ -82,9 +129,111 @@ func newCmdCertsUtility(out io.Writer) *cobra.Command {
 	cmd.AddCommand(newCmdCertsRenewal(out))
 	cmd.AddCommand(newCmdCertsExpiration(out, constants.KubernetesDir))
 	cmd.AddCommand(NewCmdCertificateKey())
+	cmd.AddCommand(newCmdGenCSR())
+	return cmd
+}
+
+// genCSRConfig is the configuration required by the gencsr command
+type genCSRConfig struct {
+	kubeadmConfigPath string
+	certDir           string
+	kubeConfigDir     string
+	kubeadmConfig     *kubeadmapi.InitConfiguration
+}
+
+func newGenCSRConfig() *genCSRConfig {
+	return &genCSRConfig{
+		kubeConfigDir: kubeadmconstants.KubernetesDir,
+	}
+}
+
+func (o *genCSRConfig) addFlagSet(flagSet *pflag.FlagSet) {
+	options.AddConfigFlag(flagSet, &o.kubeadmConfigPath)
+	options.AddCertificateDirFlag(flagSet, &o.certDir)
+	options.AddKubeConfigDirFlag(flagSet, &o.kubeConfigDir)
+}
+
+// finalize merges command line flag values into kubeadm's config.
+// Loads Kubeadm config from a file (if present)
+// else use dynamically generated default config.
+// This configuration contains the DNS names and IP addresses which
+// are encoded in the control-plane CSRs.
+func (o *genCSRConfig) finalize() (err error) {
+	o.kubeadmConfig, err = configutil.LoadOrDefaultInitConfiguration(
+		o.kubeadmConfigPath,
+		&kubeadmapiv1beta2.InitConfiguration{},
+		&kubeadmapiv1beta2.ClusterConfiguration{},
+	)
+	if err != nil {
+		return err
+	}
+	// --cert-dir takes priority over kubeadm config if set.
+	if o.certDir != "" {
+		o.kubeadmConfig.CertificatesDir = o.certDir
+	}
+	return nil
+}
+
+// newCmdGenCSR returns cobra.Command for generating keys and CSRs
+func newCmdGenCSR() *cobra.Command {
+	config := newGenCSRConfig()
+
+	cmd := &cobra.Command{
+		Use:     "generate-csr",
+		Short:   "Generate keys and certificate signing requests",
+		Long:    generateCSRLongDesc,
+		Example: generateCSRExample,
+		Args:    cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := config.finalize()
+			if err != nil {
+				return err
+			}
+			return runGenCSR(config)
+		},
+	}
+	config.addFlagSet(cmd.Flags())
 	return cmd
 }
 
+// runGenCSR contains the logic of the generate-csr sub-command after all
+// command line flags have been processed.
+func runGenCSR(config *genCSRConfig) error {
+	certDir := config.kubeadmConfig.CertificatesDir
+	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
+	if err != nil {
+		return err
+	}
+	for caCertificate, leafCertificates := range certTree {
+		klog.V(1).Infoln("[generate-csr] Generating private keys and CSRs for leaf certificates of", caCertificate.Name)
```

Comment:
should be `Infof(....%q, caCertificate.Name`), in kubeadm we mostly use Infof (or Printf from stdlib), instead of the multi-arg concat that Infoln (Println) has.

Reply:
Ah, I see that other commands use `fmt.Printf` for progress messages. I've removed all that logged output. It was unnecessary.